### PR TITLE
APS-1413 Assign keyworker

### PR DESCRIPTION
**Pull request by bobmeredith**
# Context

https://dsdmoj.atlassian.net/browse/APS-1413
This feature is available only for users with the `cas1_space_booking_record_keyworker` pernission
# Changes in this PR

Adds the ability to assign a keyworker to a placement

## Screenshots of UI changes

Placement page with open actions menu
image

Assign keyworker page
image

Error page - keyworker not selected
image

Placement page with notification and assigned keyworker
image

### CODE REVIEW

**Review by froddd — COMMENTED** [review]
Looks good, though a few suggestions I'm happy to discuss if needed!

File: integration_tests/pages/manage/placements/keyworker.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    ])
+  }
+
+  shouldShowError(): void {
+    cy.get('.govuk-error-summary__list').should('contain', 'You must select a keyworker')
+  }
+
+  completeForm(keyworkerName: string): void {
+    this.getSelectInputByIdAndSelectAnEntry('keyworkerId', keyworkerName)
+  }
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply integration_tests/pages/manage/placements/keyworker.ts:25]: The keyworker we want to assign is not the one in the placement, but one of the staffmembers from the select.
I can remove from the completeForm parameters but it does mean passing in mocked list of staff.

File: server/paths/manage.ts [hunk 2]
```diff
@@ -53,7 +53,7 @@ const deprecated = {
 const managePath = path('/manage')
 const premisesPath = managePath.path('premises')
 const singlePremisesPath = premisesPath.path(':premisesId')
-const placementPath = singlePremisesPath.path('placements/:placementId')
+const singlePlacementPath = singlePremisesPath.path('placements/:placementId')
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/paths/manage.ts:56]: Oh, I see. I was aiming for consistency, but it makes sense that premises is a special case.

File: server/utils/placements/index.test.ts [hunk 3]
```diff
@@ -22,93 +23,99 @@ describe('placementUtils', () => {
...
-          text: 'Record arrival',
-        },
-      ])
-      expect(actions(placement, { permissions: ['cas1_space_booking_record_departure'] } as UserDetails)).toEqual([
-        { classes: 'govuk-button--secondary', href: '', text: 'Record departure' },
-      ])
-      expect(actions(placement, { permissions: ['cas1_space_booking_record_keyworker'] } as UserDetails)).toEqual([
-        { classes: 'govuk-button--secondary', href: '', text: 'Allocate keyworker' },
-      ])
+    it('should require correct permissions for arrival and keyworker', () => {
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/utils/placements/index.test.ts:87]: Yes it should - thanks

File: server/controllers/manage/premises/placements/keyworkerController.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        ...userInput,
+      })
+    }
+  }
+
+  assign(): RequestHandler {
+    return async (req: Request, res: Response) => {
+      const { premisesId, placementId } = req.params
+      try {
+        const { keyworkerId } = req.body
```
[the remark this answers is not in the document]
Comment (bobmeredith) [reply server/controllers/manage/premises/placements/keyworkerController.ts:39]: Fair enough

File: server/utils/placements/index.ts [hunk 5]
```diff
@@ -77,6 +64,12 @@ export const getBackLink = (referrer: string, premisesId: string): string => {
   return paths.premises.show({ premisesId })
 }
 
+const summaryRow = (key: string, value: string, excludeRow = false): SummaryListItem =>
+  value && {
+    key: excludeRow ? null : textValue(key),
+    value: textValue(value),
+  }
+
```
Comment (froddd) [inline server/utils/placements/index.ts:72]: If this now checks the value is truthy, there should be no need for `excludeRow`, and the uses below shouldn't need the third argument either

**Review by froddd — APPROVED** [review]
A bit more cleanup needed on the `summaryRow` method, other than that good to go!

**Pull request merged**